internal/system/agent/services.go

```diff
@@ -17,165 +17,78 @@ package agent
+	rsp := make(map[string]interface{})
+	err := json.Unmarshal([]byte(response), &rsp)
+	if err != nil {
+		LoggingClient.Error("error unmarshalling response from JSON: %v", err.Error())
+	}
+	return rsp
+}
 
-func InvokeOperation(action string, services []string) error {
+func InvokeOperation(operation string, serviceNames []string) (interface{}, error) {
+	var result []interface{}
 
 	// Loop through requested operation, along with respectively-supplied parameters.
-	for _, service := range services {
-		LoggingClient.Info("invoking operation")
-
-		if !IsKnownServiceKey(service) {
-			LoggingClient.Warn(fmt.Sprintf("unknown service %s during invocation", service))
-		}
-
-		switch action {
-
-		case START:
-			if starter, ok := executorClient.(interfaces.ServiceStarter); ok {
-				err := starter.Start(service)
-				if err != nil {
-					LoggingClient.Error("error starting service")
-					return err
-				}
-			} else {
-				err := fmt.Errorf("operation not supported with specified executor")
-				LoggingClient.Error(err.Error())
-				return err
-			}
-
-		case STOP:
-			if stopper, ok := executorClient.(interfaces.ServiceStopper); ok {
-				err := stopper.Stop(service)
-				if err != nil {
-					LoggingClient.Error("error stopping service")
-					return err
-				}
-			} else {
-				err := fmt.Errorf("operation not supported with specified executor")
-				LoggingClient.Error(err.Error())
-				return err
-			}
-
-		case RESTART:
-			if restarter, ok := executorClient.(interfaces.ServiceRestarter); ok {
-				err := restarter.Restart(service)
-				if err != nil {
-					LoggingClient.Error("error restarting service")
-					return err
-				}
-			} else {
-				err := fmt.Errorf("operation not supported with specified executor")
-				LoggingClient.Error(err.Error())
-				return err
-			}
+	for _, service := range serviceNames {
+		out, err := OperationViaExecutor(service, operation)
+		if err != nil {
+			return nil, err
 		}
+		result = append(result, processResponse(out))
 	}
-	return nil
+	return result, nil
 }
 
-func getConfig(services []string, ctx context.Context) (ConfigRespMap, error) {
-
-	c := ConfigRespMap{}
-	c.Configuration = map[string]interface{}{}
+func InvokeMetrics(services []string, ctx context.Context) (interface{}, error) {
+	var result []interface{}
 
 	// Loop through requested actions, along with (any) respectively-supplied parameters.
 	for _, service := range services {
+		LoggingClient.Debug("invoke metrics")
 
-		// Check whether SMA does _not_ know of ServiceKey ("service") as being one for one of its ready-made list of clients.
-		if !IsKnownServiceKey(service) {
-			LoggingClient.Info(fmt.Sprintf("service %s not known to SMA as being in the ready-made list of clients", service))
-
-			// Service unknown to SMA, so ask the Registry whether `service` is available.
-			err := registryClient.IsServiceAvailable(service)
+		switch Configuration.MetricsMechanism {
+		case metricsOptionViaDirectService:
```

Comment:
IMHO, I think the type cast with different interfaces is a cleaner way to handle different implementations, but since I don't have time to do an indepth review I'll let other reviewers decide

Reply:
If we were doing DI, I'd agree.  Even then, you'd need to instantiate a concrete instance of the contract based on a configuration setting.  This particular implementation simplifies this without adding the overhead of an additional abstraction.